src/coreComponents/linearAlgebra/solvers/PreconditionerBlockJacobi.hpp

```diff
@@ -0,0 +1,177 @@
+    m_blockDiag->open();
+
+    array1d< globalIndex > idxBlk( m_blockSize );
+    array2d< real64 > values( m_blockSize, m_blockSize );
+    array2d< real64 > valuesInv( m_blockSize, m_blockSize );
+    for( globalIndex i = mat.ilower(); i < mat.iupper(); i+=m_blockSize )
+    {
+      values.setValues< serialPolicy >( 0.0 );
+      for( localIndex j = 0; j < m_blockSize; ++j )
+      {
+        globalIndex const iRow = i + LvArray::integerConversion< globalIndex >( j );
+        idxBlk[j] = iRow;
+        localIndex const rowLength = mat.globalRowLength( iRow );
+        array1d< globalIndex > cols( rowLength );
+        array1d< real64 > vals( rowLength );
+        mat.getRowCopy( iRow, cols, vals );
+        for( localIndex k = 0; k < rowLength; ++k )
+        {
+          localIndex const jCol = LvArray::integerConversion< localIndex >( cols[k]-i );
+          if( cols[k] >= i && cols[k] < i+LvArray::integerConversion< globalIndex >( m_blockSize ) )
+          {
+            values( j, jCol ) = vals[k];
+          }
+        }
+      }
+      BlasLapackLA::matrixInverse( values, valuesInv );
+      m_blockDiag->insert( idxBlk, idxBlk, valuesInv );
+    }
+    m_blockDiag->close();
+  }
+
+  /**
+   * @brief Compute the preconditioner from a matrix
+   * @param mat the matrix to precondition
+   * @param dofManager the Degree-of-Freedom manager associated with matrix
+   */
+  virtual void compute( Matrix const & mat,
+                        DofManager const & dofManager ) override
+  {
+    GEOSX_UNUSED_VAR( dofManager );
+    compute( mat );
+  }
+
+  /**
+   * @brief Clean up the preconditioner setup.
+   *
+   * Releases memory used and allows the matrix to be deleted cleanly.
+   * This method should be called before the matrix used to compute the preconditioner
+   * goes out of scope or is re-created. Some implementations require the matrix
+   * to outlive the preconditioner (for example, Trilinos/ML may crash the program if
+   * deleted after the matrix).
+   *
+   * @note Should be properly overridden in derived classes, which may call this method.
+   */
+  virtual void clear() override
+  {
+    m_blockDiag->reset();
+  }
+
+  /**
+   * @brief Apply operator to a vector.
+   *
+   * @param src Input vector (src).
+   * @param dst Output vector (dst).
+   */
+  virtual void apply( Vector const & src,
+                      Vector & dst ) const override
+  {
+    GEOSX_LAI_ASSERT_EQ( this->numGlobalRows(), dst.globalSize() );
+    GEOSX_LAI_ASSERT_EQ( this->numGlobalCols(), src.globalSize() );
+
+    m_blockDiag->apply( src, dst );
+  }
+
+  /**
+   * @brief Access the preconditioner in matrix form
+   * @return reference to the preconditioner matrix
+   */
+  virtual Matrix const & preconditionerMatrix() const override
+  {
+    return *m_blockDiag;
+  }
+
+private:
+
+  /// The preconditioner matrix
+  Matrix * m_blockDiag;
```

Comment:
Same as for the vector case (see above). In the matrix case, the issue arises here:
https://github.com/GEOSX/GEOSX/blob/29f3572a04efcb164ca80bb29112467d051becb2/src/coreComponents/linearAlgebra/interfaces/hypre/HypreMatrix.cpp#L61-L133

because the matrix is not ready.

Reply:
As for `PreconditionerJacobi.hpp`.